### PR TITLE
PIM-4031: Doctrine DEFERRED_EXPLICIT for product persistence

### PR DESCRIPTION
**Pull request by nidup**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | Y
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | Running http://ci.akeneo.com/job/pim-pr-behat-mongodb/696/ http://ci.akeneo.com/job/pim-pr-behat/2210
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | TODO
| Fixed tickets        | PIM-4031
| DB schema updated?   |
| Migration script?    |
| Doc PR               |

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/MongoDBODM/MetricBaseValuesSubscriber.php [hunk 1]
```diff
@@ -75,12 +78,30 @@ public function prePersist(LifecycleEventArgs $args)
...
+     * @param ProductInterface $product
+     */
+    protected function convertMetricValues(DocumentManager $dm, ProductInterface $product)
+    {
+        foreach ($product->getValues() as $value) {
+            $metric = $value->getData();
+            if ($metric instanceof MetricInterface && $metric->getUnit()) {
+                $this->createMetricBaseValues($metric);
+                if ($metric->getId() !== null) {
+                    $metadata = $dm->getClassMetadata(get_class($metric));
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/MongoDBODM/MetricBaseValuesSubscriber.php]: classutils::getclass

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/MongoDBODM/MetricBaseValuesSubscriber.php [hunk 2]
```diff
@@ -75,12 +78,30 @@ public function prePersist(LifecycleEventArgs $args)
...
+     * @param DocumentManager  $dm
+     * @param ProductInterface $product
+     */
+    protected function convertMetricValues(DocumentManager $dm, ProductInterface $product)
+    {
+        foreach ($product->getValues() as $value) {
+            $metric = $value->getData();
+            if ($metric instanceof MetricInterface && $metric->getUnit()) {
+                $this->createMetricBaseValues($metric);
+                if ($metric->getId() !== null) {
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/MongoDBODM/MetricBaseValuesSubscriber.php]: yoda

File: src/Pim/Bundle/CatalogBundle/Manager/ProductManager.php [hunk 3]
```diff
@@ -211,6 +211,18 @@ public function createProduct()
     }
 
     /**
+     * Create a product value
+     *
+     * @return \Pim\Bundle\CatalogBundle\Model\ProductValueInterface
+     *
+     * @deprecated will be removed in 1.5, please use ProductBuilderInterface::createProductValue();
+     */
+    public function createProductValue(AttributeInterface $attribute)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Manager/ProductManager.php]: why ? u don't use it in this pr
or I missed sthg maybe

Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/ProductManager.php]: I messed up the EE via the previous removing of this method, so I re-added it ;)

**Pull request merged**